custom/ewsghana/reports/specific_reports/stock_status_report.py

```diff
@@ -382,6 +339,156 @@ class StockStatus(MultiReport):
+            case_id__in=stockout_table_supply_points
+        )
+        product_case_with_stock = defaultdict(set)
+        product_case_without_stock = defaultdict(set)
+
+        months_of_stock = defaultdict(lambda: defaultdict(dict))
+        stock_state_map = {
+            (stock_state.case_id, stock_state.product_id):
+            stock_state.get_monthly_consumption() if stock_state.daily_consumption else None
+            for stock_state in stock_states
+        }
+
+        current_mos_locations = set(self.get_supply_points.values_list('supply_point_id', flat=True))
+        stockouts = defaultdict(set)
+        for (case_id, product_id, date, stock_on_hand) in transactions:
+            if stock_on_hand > 0:
+                product_case_with_stock[product_id].add(case_id)
+                if case_id in current_mos_locations:
+                    stock_state_dict = stock_state_map.get((case_id, product_id))
+                    if stock_state_dict:
+                        months_of_stock[case_id][product_id] = stock_on_hand / stock_state_dict
+                    else:
+                        months_of_stock[case_id][product_id] = None
+            else:
+                product_case_without_stock[product_id].add(case_id)
+                if case_id in current_mos_locations:
+                    stockouts[case_id].add(product_id)
+                    months_of_stock[case_id][product_id] = 0
+            grouped_by_case[case_id].add(product_id)
+
+        for case_id in current_mos_locations:
+            if case_id not in months_of_stock:
+                for product in unique_products:
+                    consumption = stock_state_map.get((case_id, product.product_id))
+                    try:
+                        stock_on_hand = StockTransaction.objects.filter(
+                            case_id=case_id,
+                            product_id=product.product_id
+                        ).latest('report__date').stock_on_hand
+                    except StockTransaction.DoesNotExist:
+                        stock_on_hand = None
+                    if stock_on_hand and consumption:
+                        months_of_stock[case_id][product.product_id] = stock_on_hand / consumption
+                    else:
+                        months_of_stock[case_id][product.product_id] = None
+
+        return {
+            'without_stock': {
+                product_id: len(case_list)
+                for product_id, case_list in product_case_without_stock.iteritems()
+            },
+            'with_stock': {
+                product_id: len(case_list)
+                for product_id, case_list in product_case_with_stock.iteritems()
+            },
+            'all': locations.count(),
+            'months_of_stock': months_of_stock,
+            'stockouts': stockouts,
+            'unique_products': unique_products,
+            'stockout_table_supply_points': stockout_table_supply_points
+        }
```

Comment:
This method is very large and does a lot of different things - it would be easier to evaluate if you split up and encapsulated chunks of it.